### ISSUE TITLE
Cumulative sum runs slowly with EXLA backend

### ISSUE THREAD

**Issue by jonatanklosko** [elixir-nx/nx#887]
This takes around 200s for me:

```elixir
Mix.install([
  {:nx, github: "elixir-nx/nx", sparse: "nx", override: true},
  {:exla, github: "elixir-nx/nx", sparse: "exla", override: true}
])

Nx.default_backend(EXLA.Backend)

x = Nx.iota({200000})
Nx.cumulative_sum(x)
```

Meanwhile, the corresponding jax script finishes immediately:

```python
import jax.numpy as jnp

x = jnp.arange(200000)
jnp.cumsum(x)
```

**Comment by polvalente** [comment]
@jonatanklosko EXLA.Backend does not implement a custom version of the `cumulative_sum` option callback, which ends up falling back to the default implementation which relies on padding the tensor so it can use `window_sum` to evaluate the cumulative sum.

It seems that the problem lies in the `window_sum` implementation given the following code which mimics the default implementation:

```elixir
# this one executes "instantly", where x is the same iota as you defined
Nx.pad(x, 0, [{Nx.size(x) - 1, 0, 0}]) 
# This one takes too long
Nx.pad(x, 0, [{Nx.size(x) - 1, 0, 0}]) |> Nx.window_sum({Nx.size(x)})
```

**Comment by josevalim** [comment]
We should check how JAX implements it. There is likely a way to skip the pad dor make it more efficient? :)

**Comment by polvalente** [comment]
From _src/numpy/lax_numpy.py, there's this comment which might shed some light: `# We want to allow XLA to fuse the pad and reduce-window operators to
  # avoid materializing the padded output.`
  
  Testing locally, what seems to be the root of the problem is indeed the padding size, as shown below:
  
  ```elixir
  {t1, r1} = :timer.tc(fn -> Nx.default_backend(EXLA.Backend); Nx.window_sum(Nx.concatenate([Nx.broadcast(0, {200000}), Nx.iota({200000})]), {200000}) end)
  {1108008,
 #Nx.Tensor<
   s64[200001]
   EXLA.Backend<host:0, 0.1733127262.4285136912.245461>
   [0, 0, 1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 66, 78, 91, 105, 120, 136, 153, 171, 190, 210, 231, 253, 276, 300, 325, 351, 378, 406, 435, 465, 496, 528, 561, 595, 630, 666, 703, 741, 780, 820, 861, 903, 946, 990, 1035, 1081, ...]
 >}
iex(3)> {t2, r2} = :timer.tc(fn -> Nx.default_backend(EXLA.Backend); Nx.window_sum(Nx.iota({200000}), {200000}, padding: 
{53780588,
 #Nx.Tensor<
   s64[200001]
   EXLA.Backend<host:0, 0.1733127262.4285136912.245463>
   [0, 0, 1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 66, 78, 91, 105, 120, 136, 153, 171, 190, 210, 231, 253, 276, 300, 325, 351, 378, 406, 435, 465, 496, 528, 561, 595, 630, 666, 703, 741, 780, 820, 861, 903, 946, 990, 1035, 1081, ...]
 >}
 # There are ~52s of difference between the first execution and the second
 iex(7)> (t2 - t1) |> div(1_000_000)
 52
 # The results are the same
 iex(11)> Nx.equal(r1, r2) |> Nx.all()
#Nx.Tensor<
  u8
  EXLA.Backend<host:0, 0.1733127262.4285136912.245466>
  1
>
  ```

**Comment by josevalim** [comment]
Using EXLA.jit is instantaneous though, so it feels like EXLA is not optimizing those callbacks accordingly:

```
iex(1)> :timer.tc(fn -> EXLA.jit(fn -> Nx.window_sum(Nx.iota({200000}), {200000}, padding: [{200000, 0}]) end).() end)
```

**Comment by josevalim** [comment]
Wait, on my machine it is always instantaneous, so I guess it is not a good place to test. :(

@polvalente, does JIT make a difference for you? If it does, then it means we are not jitting optional callbacks accordingly. In any case, whatever JAX does to avoid the padding is likely a good idea.

**Comment by polvalente** [comment]
@josevalim the code you provided is as slow as the one I ran before on my machine. Perhaps your CPU limits are better? It's not linear though, so adding a zero or two to your padding and window size might show the same problem in your machine.

**Comment by polvalente** [comment]
Update: after removing some debug inspects, JIT makes it run instantaneously

**Comment by jonatanklosko** [comment]
> Update: after removing some debug inspects, JIT makes it run instantaneously

Which one? For me all of the above run slow.

**Comment by jonatanklosko** [comment]
I did more investigation and the pad + window sum implementation is exactly as slow as ours.

```python
import jax
import jax.numpy as jnp

def _cumsum(x):
    size = x.shape[0]
    x = jax.lax.pad(x, 0, [(size - 1, 0, 0)])
    return jax.lax.reduce_window(x, 0, jax.lax.add, (size,), (1,), 'VALID')

cumsum = jax.jit(_cumsum)

x = jnp.arange(200000)
cumsum(x)
```

It turns out that jax only uses this implementation on TPU ([ref](https://github.com/google/jax/blob/da24b99d308141cb8c477a10ad9ac6239e0643f3/jax/_src/lax/control_flow/loops.py#L1957)), where it's further optimised by the compiler ([ref](https://github.com/google/jax/blob/da24b99d308141cb8c477a10ad9ac6239e0643f3/jax/_src/lax/control_flow/loops.py#L1915-L1917)). Otherwise it uses [associative scan](https://github.com/google/jax/blob/da24b99d308141cb8c477a10ad9ac6239e0643f3/jax/_src/lax/control_flow/loops.py#L1742) ([ref](https://github.com/google/jax/blob/da24b99d308141cb8c477a10ad9ac6239e0643f3/jax/_src/lax/control_flow/loops.py#L1953)), reducing it from O(N^2) to O(n log n) ([ref](https://github.com/google/jax/pull/2596)).

**Comment by polvalente** [comment]
It seems like we can implement the optional callback for EXLA with some minor modifications to the EXLA compiler (and Torchx without any mods)

**Comment by jonatanklosko** [comment]
We likely want to just optimise the default implementation, rather EXLA.

I implemented cumulative sum using associative scan and now the same example runs for 0.6s using the binary backend. With EXLA it takes ~1s to JIT and then runs under 0.3ms!

<details>
<summary>Code</summary>

```elixir
Mix.install([
  {:nx, github: "elixir-nx/nx", sparse: "nx", override: true},
  {:exla, github: "elixir-nx/nx", sparse: "exla", override: true}
])

defmodule Scan do
  import Nx.Defn

  defn cumulative_sum(x, opts \\ []) do
    opts = keyword!(opts, axis: 0)
    associative_scan(x, &Nx.add/2, axis: opts[:axis])
  end

  defn associative_scan(x, fun, opts) do
    opts = keyword!(opts, axis: 0)
    axis = opts[:axis]

    axis_size = Nx.axis_size(x, axis)

    if axis_size < 2 do
      x
    else
      even = Nx.slice_along_axis(x, 0, axis_size - 1, axis: axis, strides: 2)
      odd = Nx.slice_along_axis(x, 1, axis_size - 1, axis: axis, strides: 2)

      reduced_pairs = fun.(odd, even)

      scanned_odd = associative_scan(reduced_pairs, fun, opts)

      cond do
        axis_size == 2 ->
          Nx.concatenate([even, reduced_pairs], axis: axis)

        rem(axis_size, 2) == 0 ->
          scanned_even =
            fun.(
              Nx.slice_along_axis(scanned_odd, 0, div(axis_size, 2) - 1, axis: axis),
              Nx.slice_along_axis(even, 1, div(axis_size, 2) - 1, axis: axis)
            )

          scanned_even =
            Nx.concatenate(
              [Nx.slice_along_axis(even, 0, 1, axis: axis), scanned_even],
              axis: axis
            )

          interleave(scanned_even, scanned_odd, axis: axis)

        true ->
          scanned_even =
            fun.(
              scanned_odd,
              Nx.slice_along_axis(x, 2, axis_size - 2, axis: axis, strides: 2)
            )

          Nx.concatenate(
            [
              Nx.slice_along_axis(x, 0, 1, axis: axis),
              interleave(scanned_odd, scanned_even, axis: axis)
            ],
            axis: axis
          )
      end
    end
  end

  # Interleaves elements from same-shaped tensors along an axis
  defnp interleave(x, y, opts \\ []) do
    opts = keyword!(opts, axis: 0)
    axis = opts[:axis]

    interleave_axis = axis + 1

    Nx.concatenate(
      [
        Nx.new_axis(x, interleave_axis),
        Nx.new_axis(y, interleave_axis)
      ],
      axis: interleave_axis
    )
    |> flatten_axis(interleave_axis)
  end

  # Merges the given axis with the preceding one
  deftransformp flatten_axis(tensor, axis) do
    shape = Nx.shape(tensor)
    new_shape = shape |> Tuple.delete_at(axis) |> put_elem(axis - 1, :auto)
    Nx.reshape(tensor, new_shape)
  end
end

Nx.default_backend(EXLA.Backend)

x = Nx.iota({200_000})

:timer.tc(fn -> EXLA.jit(&Scan.cumulative_sum/1).(x) end) |> elem(0) |> IO.inspect(label: "jit 1")
:timer.tc(fn -> EXLA.jit(&Scan.cumulative_sum/1).(x) end) |> elem(0) |> IO.inspect(label: "jit 2")
```

</details>

**Comment by polvalente** [comment]
That leave me with only one thing to say: :exploding_head:
